respy/method_of_simulated_moments.py

```diff
@@ -0,0 +1,140 @@
+"""Estimate models with the method of simulated moments (MSM).
+
+The method of simulated moments is developed by [1], [2], and [3] and an estimation
+technique where the distance between the moments of the actual data and the moments
+implied by the model parameters is minimized.
+
+References
+----------
+
+.. [1] McFadden, D. (1989). A method of simulated moments for estimation of discrete
+       response models without numerical integration. Econometrica: Journal of the
+       Econometric Society, 995-1026.
+.. [2] Lee, B. S., & Ingram, B. F. (1991). Simulation estimation of time-series models.
+       Journal of Econometrics, 47(2-3), 197-205.
+.. [3] Duffie, D., & Singleton, K. (1993). Simulated Moments Estimation of Markov Models
+       of Asset Prices. Econometrica, 61(4), 929-952.
+
+"""
+import functools
+
+import numpy as np
+import scipy as sc
+import pandas as pd
+
+from respy.pre_processing.model_processing import process_params_and_options
+from respy.simulate import get_simulate_func
+
+
+def get_msm_func(
+    params,
+    options,
+    moments,
+    calc_moments,
+    weighting_matrix=None,
+    replacement=0,
+    all_dims=False,
+):
+    """Get the criterion function for estimation with MSM.
+
+    Return a version of the :func:`msm` function where all arguments except the
+    parameter vector are fixed with :func:`functools.partial`. Thus the function can be
+    directly passed into an optimizer or a function for taking numerical derivatives.
+
+    Parameters
+    ----------
+    params : pandas.DataFrame
+        DataFrame containing model parameters.
+    options : dict
+        Dictionary containing model options.
+    moments : np.ndarray
+        Array with shape (n_moments,) containing the moments of the actual data.
+    calc_moments : callable
+        Function to compute the moments from the simulated data.
+    weighting_matrix : np.ndarray
+        Array with shape (n_moments, n_moments) which is used to weight the squared
+        errors of moments.
+    replacement : callable or int
+        If int then all missing moments of the evaluation with the current moments
+         are set to this int. Else callable that maps missing indices into
+          replacement int.
+    all_dims: boolean
+        If set to false objective is returned. If set to True weighted array
+        of component deviations will be returned.
+
+    Returns
+    -------
+    msm_function : :func:`msm`
+        Criterion function where all arguments except the parameter vector are set.
+
+    """
+    optim_paras, options = process_params_and_options(params, options)
+
+    simulate = get_simulate_func(params, options)
+
+    if weighting_matrix is None:
```

Comment:
I am more in favor of making it a required input. If someone wants to get away with the identity matrix, it is just a one-liner.